docs/proposals/manifest-hydrator.md

```diff
@@ -0,0 +1,538 @@
+
+Equipped with this information, the developer can perform the exact same steps as Argo CD and be confident that their dry manifest changes will produce the desired output.
+
+Ensuring that hydration is deterministic assures the developer that the output for a given dry state will be the same next week as it is today.
+
+###### Avoiding Esoteric Behavior
+
+We should avoid the developer needing to know Argo CD-specific behavior in order to reproduce hydration. Tools like Helm, Kustimize, etc. have excellent public-facing documentation which the developer should be able to take advantage of without needing to know quirks of Argo CD.
+
+##### Reliable Previews
+
+Deterministic hydration output allows Argo CD to produce a reliable change preview when a developer proposes a change to the dry manifests via a PR.
+
+If output is not deterministic, then a preview generated today might not be valid/correct a week, day, or even hour later. Non-determinism makes it so that developers can't trust that the change they review will be the change actually applied.
+
+### Non-Goals
+
+1) Implementing a change promotion system
+
+## Open Questions
+
+* The `sourceHydrator` field is mutually exclusive with the `source` and the `sources` field. Should we throw an error if they're both configured, or should we just pick one and ignore the others?
+* How will/should this feature relate to the image updater? Is there an opportunity to share code, since both tools involve pushing to git?
+* Should we enforce a naming convention for hydrated manifest branches, e.g. `argo/...`? This would make it easier to recommend branch protection rules, for example, only allow pushes to `argo/*` from the argo bot.
+* Should we enforce setting a `sourceHydrator.syncSource.path` to something besides `.`? Setting a path makes it easier to add/remove other apps later if desired.
+
+## Proposal
+
+Today, Argo CD watches one or more git repositories (configured in the `spec.source` or `spec.sources` field). When a new commit appears, Argo CD updates the desired state by rendering the manifests with the configured manifest hydration tool. If auto-sync is enabled, Argo CD applies the new manifests to the cluster.
+
+With the introduction of this change, Argo CD will watch two revisions in the same git repository: the first is the "dry source", i.e. the git repo/revision where the un-rendered manifests reside, and the second is the "hydrated source," where the rendered manifests are places and retrieved for syncing to the cluster.
+
+### New `spec.sourceHydrator` Application Field
+
+A `sourceHydrator` field will be added to the Argo CD Application spec:
+
+```yaml
+apiVersion: argoproj.io/v1alpha1
+kind: Application
+metadata:
+  name: example
+spec:
+  # The sourceHydrator field is mutually-exclusive with `source` and with `sources`. If this field is configured, we 
+  # should either throw an error or ignore the other two.
+  sourceHydrator:
```

Comment:
@crenshaw-dev @jessesuen @christianh814 
Now thinking about it, perhaps the way it is added to the spec could be streamlined. I am thinking of something like the following:
```yaml
spec:
  project: default
  source:
    repoURL: https://github.com/argoproj/argocd-example-apps
    targetRevision: main
    path: environments/e2e
    hydration:
      drySource:
        targetBranch: environments/e2e
        path: .
      hydrateTo:
        targetBranch: environments/e2e
```
I don't know if it would work, but it would
- be an easy migration for users currently using the dry pattern. They only need to add the `hydration` config
- Fit the current "rendering" tools pattern
- Smaller impact on CRD. No need to check for conflict with `source` or `sources`.
- Explicitly no support for multi-sources this was, but it could be added with low CRD changes after.
- Benefit of the current `manifest-generate-paths` annotation.

Now the hydrator becomes more of a CMP on steroids, but it cannot be implemented as a CMP because it will need to watch for changes on `source.hydration.drySource.targetBranch` so I think we get a big benefit from implementing this as a built-in tool.

Reply:
Reading this and then relooking at the spec, I think your (@agaudreault) draft spec thoughts makes the most sense to me.

```yaml
spec:
  source:
    repoURL: https://github.com/argoproj/argocd-example-apps
    targetRevision: main
    path: environments/e2e
    hydration:
      syncSource:
        targetBranch: environments/e2e
        path: .
```

This way, my gut reaction, feels more "natural" to me on first read. It streamlines it and reads "take this source, hydrate it, and have Argo CD sync it from the hydrated branch"

Reply:
I think there are two suggested changes here:

1. Use repo-server to initiate the `git push` (delegated to a separate commit-server component)
2. Nest hydrator config under the `source` field

To answer the first, I think there are a few downsides to a side-effect based write-back system in the repo-server. But I think the deal-breaker is the lack of a way to aggregate hydration across multiple apps. The current hydrator implementation (in the app controller) finds all apps which hydrate to a specific repo/branch pair, collects their hydrated manifests, and pushes those manifests in a single git commit. The repo-server currently operates only on a per-app basis. To aggregate apps for single-commit hydration, we'd have to make the repo-server multi-app-aware.

There are other downsides. If aggregation doesn't sound like a deal-breaker, we can dive into other issues and see if they're surmountable.

To answer the second suggestion: I like the simplicity, especially of the migration experience. But I think there are some downsides here, too. I'll split my concerns into two categories: UX and backend.

### UX Concerns

#### 1. `spec.source.chart` isn't obviously invalid

The hydrator currently only supports git sources. The nested `hydrator` API makes it look as if maybe we support Helm sources as well.

```yaml
source:
  repoURL: https://helm.example.com/my-chart
  chart: my-chart # at a glance, looks like this should work, right?
  hydration:
    syncSource: 
```

It's possible that we'll support Helm charts in the future (though I kind of doubt it, since deterministic hydration is a key aspect of this feature, and the only possible deterministic hydration of a Helm chart is one in which there's no user input).

#### 2. Lack of parallelism with `helm`, `kustomize`, and `directory`

As the user reads an example application.yaml, they'll encounter the `helm`, `kustomize`, and `directory` fields under `spec.source`. In the case of each of those fields, the behavior they provide is basically "pass some parameters to the hydration tool." If we add a `hydration` field at the same level, we break that pattern. Now instead of just configuring inputs for a build tool, we're configuring an entire write-back system. It's just a surprising deviation from the current pattern.

#### 3. Not obviously mutually-exclusive with `helm`, `kustomize`, and `directory`

One nice thing about a new top-level `sourceHydrator` field is that it's obvious how new and different it is from the `source` API. If we use the `source` API, the user might think that this is possible:

```yaml
source:
  hydration: # all my hydrator config
  kustomize: # my Kustomize overrides
```

We have a similar problem today where folks want to configure `helm` and `kustomize` inputs at the same time. But in that case it's a bit more intuitive why they might be mutually-exclusive: you either run `helm` or you run `kustomize`. In the case of the hydrator, a user might reasonably wonder "why can't I run the hydrator with some Kustomize inputs"? There's a good answer having to do with deterministic hydration, but it's not obvious from the API that we're enforcing that opinion (until you try it and get a runtime failure).

### Backend Concerns

I'm generally strongly opposed to letting the backend dictate the API. The user experience should be the ultimate concern.

But in this case, I think there are good reasons to believe having a top-level `sourceHydrator` field is a better UX. So maybe backend concerns can serve as a tie-breaker.

#### 1. `spec.source` is basically a repo-server API request body

The app-controller and repo-server basically split concerns today like this: `spec.source` is the repo-server's concern, and everything else is an app-controller concern. So the contents of `spec.source` are literally an API request body to the repo-server.

If hydration is not performed by the repo-server (reasons why given above), then it doesn't really make sense to send the hydration config to the repo-server. And doing so could entice maintainers to do silly things in the future (for example, to render manifests differently depending on their destination repoURL/branch/path).

We could have the app-controller clear out the `hydration` field before sending the `source` field to repo-server, but it would be strange code to write/test/maintain.

#### 2. Multi-source would be weird, or types would be duplicated

Today `spec.source` is unmarshaled to a struct of type `ApplicationSource`. `spec.sources` is type `[]ApplicationSource`. It's unclear what it would mean to have multiple sources in a single app hydrating to different destinations. Conceptually it seems possible (hydrate to all, then coalesce the manifests into a single document for syncing). But the source hydrator as currently proposed intentionally does not work with multi-source apps.

So we'd either have to duplicate the `ApplicationSource` struct into a new type `ApplicationSingleSource` to add the `hydration` field, or we'd have to let the `hydration` field appear in multiple sources and somehow handle it there (either by implementing multi-source hydration, throwing a runtime error, or ignoring the field).

tl;dr - I think app-controller is better than repo-server to coordinate commit-server API calls, and I think a new top-level field is justified despite small UX downsides.

Reply:
+1 to @crenshaw-dev around Multi-Source applications. If the `hydrator` field is added to the `ApplicationSource` spec, the `sources` field would also take in the same change. There are multiple areas in our codebase where we pass in just a single source and do not have a mechanism to identify if the application is a single source app or a multi-source app.

Also, duplicating the `ApplicationSource` struct into a new type `ApplicationSingleSource` type might be a breaking change for folks relying on the current `ApplicationSource` format.